libsignal-service/src/sealed_session_cipher.rs

```diff
@@ -0,0 +1,1070 @@
+    ) -> Result<Self, SealedSessionError> {
+        use crate::proto::sender_certificate::Certificate;
+        use prost::Message;
+        match (wrapper.signature, wrapper.certificate) {
+            (Some(signature), Some(certificate)) => {
+                let Certificate {
+                    sender_e164,
+                    sender_uuid,
+                    sender_device,
+                    expires,
+                    identity_key,
+                    signer,
+                } = Message::decode(&certificate[..])?;
+                match (sender_device, expires, identity_key, signer) {
+                    (
+                        Some(sender_device_id),
+                        Some(expires),
+                        Some(identity_key),
+                        Some(signer),
+                    ) => {
+                        if sender_e164.is_none() && sender_uuid.is_none() {
+                            return Err(SealedSessionError::InvalidCertificate);
+                        }
+
+                        Ok(Self {
+                            signer: ServerCertificate::try_from(
+                                &context, signer,
+                            )?,
+                            key: PublicKey::decode_point(
+                                &context,
+                                &identity_key,
+                            )?,
+                            sender_e164,
+                            sender_uuid,
+                            sender_device_id: sender_device_id as i32,
+                            expiration: expires,
+                            certificate,
+                            signature,
+                        })
+                    }
+                    _ => Err(SealedSessionError::InvalidCertificate),
+                }
+            }
+            _ => Err(SealedSessionError::InvalidCertificate),
+        }
+    }
+}
+
+impl ServerCertificate {
+    fn try_from(
+        context: &Context,
+        wrapper: crate::proto::ServerCertificate,
+    ) -> Result<Self, SealedSessionError> {
+        use crate::proto::server_certificate;
+        use prost::Message;
+
+        match (wrapper.certificate, wrapper.signature) {
+            (Some(certificate), Some(signature)) => {
+                let server_certificate: server_certificate::Certificate =
+                    Message::decode(&certificate[..])?;
+                match (server_certificate.id, server_certificate.key) {
+                    (Some(id), Some(key)) => Ok(Self {
+                        key_id: id,
+                        key: PublicKey::decode_point(context, &key)?,
+                        certificate,
+                        signature,
+                    }),
+                    _ => Err(SealedSessionError::InvalidCertificate),
+                }
+            }
+            _ => Err(SealedSessionError::InvalidCertificate),
+        }
+    }
+}
+
+// impl TryInto<Vec<u8>> for ServerCertificate {
```

Comment:
Commented code, should this go, or stay, or be rewritten, or be put in an issue?